src/mpid/ch4/netmod/include/netmod_am_fallback_send.h

```diff
@@ -14,7 +14,12 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_mpi_send(const void *buf,
                                                MPIR_Comm * comm, int context_offset,
                                                MPIDI_av_entry_t * addr, MPIR_Request ** request)
 {
-    return MPIDIG_mpi_send(buf, count, datatype, rank, tag, comm, context_offset, addr, request);
+    int mpi_errno;
+    int protocol = MPIDIG_AM_PROTOCOL__EAGER;
+
+    protocol = MPIDI_NM_am_choose_protocol(buf, count, datatype, 0, MPIDIG_SEND);
```

Comment:
This is one place where the double meaning of `MPIDIG_SEND` could confusing. Here it means this is for a regular send (not indicating EAGER protocol).

Reply:
`MPIDIG_SEND` should really be internal to the generic layer. I don't think the netmod should use it.

Reply:
The NM/SHM just pass this variable to the MPIDIG_ function that calculates whether the given `data_sz` and message type is within the eager limit provided by NM/SHM.

This is a somewhat unfortunate design because we want to move the decision on protocol before calling the `MPIDIG_mpi_` functions.  Three pieces of info are needed for this: `data_sz`, MPIDIG level `am_hdr_sz` (which is variable depends on operation type) and the NM/SHM eager limit (queried from NM/SHM function).

Reply:
We'll need potentially three distinct namespaces: 

* one for mpi-level message types, e.g. SEND, PUT, ... question: do we need distinguish them? I.e. do we expect a byte-playload from SEND or PUT be sent differently?

* one for the am handler_id. This is what current MPIDIG_SEND is. It doesn't need to be tied to anything other than callbacks. So if `MPIDIG_SEND` will be split into `EAGER|RNDV|SSEND`, then we probably don't need `MPIDIG_SEND`.

* one for the byte-payload transport mechanism. This is `EAGER|PIPELINE|RDMA`. Note that this type is orthogonal to handler_id. I.e. for every protocol_id that has a payload, well multiplex them, e.g. `PUT_DATA+EAGER`, `PUT_DATA+PIPELINE`, `PUT_DATA+RDMA`.

So to avoid confusions, we need clarify on the terms. If we call a combination/series of handler_ids as protocol -- the conventional usage -- then we should not call the byte transfer method also protocol. How about "transport method"?

Another term we need clarify is "payload" and "data". There is mpi-level data that @pavanbalaji refer to as payload; then there is transport level "data" that is after "headers". I used to call the opposite (upper level "data" and lower layer "payload"). Either way is fine (just need getting used to), but we need clarify up-front now.